update: check SignatureMediaType in notation.Verify

patrickzheng200 opened the pull request:
In this PR:
1. Created RemoteVerifyOptions for notation.Verify(), and therefore, VerifyOptions is now only for verifier.Verify().
2. In notation.Verify(), using signature media type fetched from registry.

Signed-off-by: Patrick Zheng <patrickzheng@microsoft.com>

JeyJeyGao commented on notation.go:
```diff
@@ -177,10 +177,19 @@ func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, op
+				continue
+			}
+			// Otherwise, use the MediaType fetched from repo
+			opts.SignatureMediaType = sigDesc.MediaType
```
A little confusing. Correct me if I'm wrong.
Suppose we have 2 sigantures and opts.SignatureMediaType is empty:
1. when processing the first one `sigCose1`:  opts.SignatureMediaType = "application/cose"
2. when processing the second one `sigJWS2`: this signature will be skipped.
Because opts.SignatureMediaType is `empty`, user don't want to skip any signature type.

patrickzheng200 replied:
Code updated.

JeyJeyGao replied:
Is it possible that the user reuses the `VerifyOptions` but the user doesn't know we changed the `opts.SignatureMediaType` which causing a issue?

priteshbandi reviewed (commented):
Why are we even taking `opts.SignatureMediaType` as input? Won't hurt multi-signature envelope support?

patrickzheng200 commented:
> Why are we even taking `opts.SignatureMediaType` as input? 
This enables users to pick their trusted signature format. Signature format that they don't trust will not be processed even if it might pass the verification process.

> Also, will checking check SignatureMediaType in notation.Verify hurt multi-signature envelope support?
No, it won't hurt multi-signature format support. 
Basically, there are 3 scenarios:
1. VerifyOptions.SignatureMediaType is empty. In this case, MediaType fetched from repo will be used.
2. VerifyOptions.SignatureMediaType is not empty but does not match MediaType fetched from repo. In this case, this certain signature will be skipped, and the program will continue to verify the next signature.
3. VerifyOptions.SignatureMediaType equals to MediaType fetched from repo. This MediaType will be used to verify the signature.
To enable both JWS and COSE, user could just leave VerifyOptions.SignatureMediaType as empty. (I updated the code to reflect this.)

shizhMSFT reviewed (approved):
LGTM

JeyJeyGao reviewed (approved):
LGTM

priteshbandi commented on notation.go:
```diff
@@ -169,6 +169,7 @@ func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, op
 	}
 	errExceededMaxVerificationLimit := ErrorVerificationFailed{Msg: fmt.Sprintf("total number of signatures associated with an artifact should be less than: %d", opts.MaxSignatureAttempts)}
 	numOfSignatureProcessed := 0
+	userSignaureMediaType := opts.SignatureMediaType
```
`opts.SignatureMediaType` should be an array to allow user to trust multiple signature formats. Also, an empty array here means they trust all.

patrickzheng200 replied:
Added unit test. As for changing it to array, IMO it's not very necessary at this point since we only support two formats. An empty user input means supporting both of them. Besides, changing VerifyOption would require Ratify to change accordingly.

priteshbandi replied:
If this is external facing method used by ratify, I would suggest changing this to array to avoid any breaking changes post RC1.

priteshbandi replied:
I am confused, either we need to take this as an array or we should completely drop this functionality and always use `mediaType` present signature artifact manifest.

priteshbandi commented:
> > Why are we even taking `opts.SignatureMediaType` as input?
> 
> This enables users to pick their trusted signature format. Signature format that they don't trust will not be processed even if it might pass the verification process.
> 
The question here is, if a user trust notary v2 for signing and verifying, why wont they trust a signature format that has been vetted by notary v2 ?

> > Also, will checking check SignatureMediaType in notation.Verify hurt multi-signature envelope support?
> 
> No, it won't hurt multi-signature format support. Basically, there are 3 scenarios:
> 
> 1. VerifyOptions.SignatureMediaType is empty. In this case, MediaType fetched from repo will be used.
> 2. VerifyOptions.SignatureMediaType is not empty but does not match MediaType fetched from repo. In this case, this certain signature will be skipped, and the program will continue to verify the next signature.
> 3. VerifyOptions.SignatureMediaType equals to MediaType fetched from repo. This MediaType will be used to verify the signature.
> 
> To enable both JWS and COSE, user could just leave VerifyOptions.SignatureMediaType as empty. (I updated the code to reflect this.)

Yes, that's an implementation detail. I am wondering why will a user want to trust one signature format but not the other. Also, if an implementation does that then it's not abiding by notartv2 spec.

patrickzheng200 commented:
> > > Why are we even taking `opts.SignatureMediaType` as input?
> > 
> > 
> > This enables users to pick their trusted signature format. Signature format that they don't trust will not be processed even if it might pass the verification process.
> 
> The question here is, if a user trust notary v2 for signing and verifying, why wont they trust a signature format that has been vetted by notary v2 ?
> 
> > > Also, will checking check SignatureMediaType in notation.Verify hurt multi-signature envelope support?
> > 
> > 
> > No, it won't hurt multi-signature format support. Basically, there are 3 scenarios:
> > 
> > 1. VerifyOptions.SignatureMediaType is empty. In this case, MediaType fetched from repo will be used.
> > 2. VerifyOptions.SignatureMediaType is not empty but does not match MediaType fetched from repo. In this case, this certain signature will be skipped, and the program will continue to verify the next signature.
> > 3. VerifyOptions.SignatureMediaType equals to MediaType fetched from repo. This MediaType will be used to verify the signature.
> > 
> > To enable both JWS and COSE, user could just leave VerifyOptions.SignatureMediaType as empty. (I updated the code to reflect this.)
> 
> Yes, that's an implementation detail. I am wondering why will a user want to trust one signature format but not the other. Also, if an implementation does that then it's not abiding by notartv2 spec.

Please don't get me wrong, I don't mean users should or need to pick one specific signature format here. Leaving the field empty is the default behavior, i.e. taking both signature formats is our default.  The purpose of this PR is not enabling users to pick a signature format. On the contrary, previous to this PR, the logic is only using opts.SignatureMediaType. This PR actually fixes this issue and takes SignatureMediaType from registry.

priteshbandi commented:
> > > > Why are we even taking `opts.SignatureMediaType` as input?
> > > 
> > > 
> > > This enables users to pick their trusted signature format. Signature format that they don't trust will not be processed even if it might pass the verification process.
> > 
> > 
> > The question here is, if a user trust notary v2 for signing and verifying, why wont they trust a signature format that has been vetted by notary v2 ?
> > > > Also, will checking check SignatureMediaType in notation.Verify hurt multi-signature envelope support?
> > > 
> > > 
> > > No, it won't hurt multi-signature format support. Basically, there are 3 scenarios:
> > > 
> > > 1. VerifyOptions.SignatureMediaType is empty. In this case, MediaType fetched from repo will be used.
> > > 2. VerifyOptions.SignatureMediaType is not empty but does not match MediaType fetched from repo. In this case, this certain signature will be skipped, and the program will continue to verify the next signature.
> > > 3. VerifyOptions.SignatureMediaType equals to MediaType fetched from repo. This MediaType will be used to verify the signature.
> > > 
> > > To enable both JWS and COSE, user could just leave VerifyOptions.SignatureMediaType as empty. (I updated the code to reflect this.)
> > 
> > 
> > Yes, that's an implementation detail. I am wondering why will a user want to trust one signature format but not the other. Also, if an implementation does that then it's not abiding by notartv2 spec.
> 
> Please don't get me wrong, I don't mean users should or need to pick one specific signature format here. Leaving the field empty is the default behavior, i.e. taking both signature formats is our default. The purpose of this PR is not enabling users to pick a signature format. On the contrary, previous to this PR, the logic is only using opts.SignatureMediaType. This PR actually fixes this issue and takes SignatureMediaType from registry.

if the purpose of this PR is not enabling users to pick a signature format, can we always use mediatype returned by signature manifest i.e. user shouldn't be able to choose?

patrickzheng200 commented:
> > > > > Why are we even taking `opts.SignatureMediaType` as input?
> > > > 
> > > > 
> > > > This enables users to pick their trusted signature format. Signature format that they don't trust will not be processed even if it might pass the verification process.
> > > 
> > > 
> > > The question here is, if a user trust notary v2 for signing and verifying, why wont they trust a signature format that has been vetted by notary v2 ?
> > > > > Also, will checking check SignatureMediaType in notation.Verify hurt multi-signature envelope support?
> > > > 
> > > > 
> > > > No, it won't hurt multi-signature format support. Basically, there are 3 scenarios:
> > > > 
> > > > 1. VerifyOptions.SignatureMediaType is empty. In this case, MediaType fetched from repo will be used.
> > > > 2. VerifyOptions.SignatureMediaType is not empty but does not match MediaType fetched from repo. In this case, this certain signature will be skipped, and the program will continue to verify the next signature.
> > > > 3. VerifyOptions.SignatureMediaType equals to MediaType fetched from repo. This MediaType will be used to verify the signature.
> > > > 
> > > > To enable both JWS and COSE, user could just leave VerifyOptions.SignatureMediaType as empty. (I updated the code to reflect this.)
> > > 
> > > 
> > > Yes, that's an implementation detail. I am wondering why will a user want to trust one signature format but not the other. Also, if an implementation does that then it's not abiding by notartv2 spec.
> > 
> > 
> > Please don't get me wrong, I don't mean users should or need to pick one specific signature format here. Leaving the field empty is the default behavior, i.e. taking both signature formats is our default. The purpose of this PR is not enabling users to pick a signature format. On the contrary, previous to this PR, the logic is only using opts.SignatureMediaType. This PR actually fixes this issue and takes SignatureMediaType from registry.
> 
> if the purpose of this PR is not enabling users to pick a signature format, can we always use mediatype returned by signature manifest i.e. user shouldn't be able to choose?

Code updated. Changed the field to an array.

shizhMSFT commented on notation.go:
```diff
@@ -94,6 +95,21 @@ type VerifyOptions struct {
 	MaxSignatureAttempts int
```
The documentation of this field needs to be updated.

shizhMSFT commented on notation.go:
```diff
@@ -156,31 +178,38 @@ func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, op
 			if err != nil {
 				return ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("unable to retrieve digital signature with digest %q associated with %q from the registry, error : %v", sigManifestDesc.Digest, artifactRef, err.Error())}
 			}
+			if len(remoteOpts.SignatureMediaTypes) != 0 && !slices.Contains(remoteOpts.SignatureMediaTypes, sigDesc.MediaType) {
+				continue
+			}
```
The performance of this check is slow since it checks the media type after fetching the blob. Could you create an issue to track it?

patrickzheng200 replied:
created an issue to track this. (#215)

shizhMSFT reviewed (approved):
LGTM

patrickzheng200 commented:
> 

@priteshbandi Oh yeah, I've updated the code. PTAL.

priteshbandi commented on notation.go:
```diff
@@ -141,12 +134,33 @@ type Verifier interface {
+	// PluginConfig is a map of plugin configs.
+	PluginConfig map[string]string
+
+	// MaxSignatureAttempts is the maximum number of signature envelopes that
+	// can be associated with the target artifact. If set to less than or equals
+	// to zero, an error will be returned.
```
```suggestion
	// MaxSignatureAttempts is the maximum number of signature envelopes that
	// will be processed for verification. If set to less than or equals
	// to zero, an error will be returned.
```

priteshbandi commented on notation.go:
```diff
@@ -156,31 +170,36 @@ func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, op
 	var verificationOutcomes []*VerificationOutcome
-	if opts.MaxSignatureAttempts <= 0 {
-		return ocispec.Descriptor{}, verificationOutcomes, ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("verifyOptions.MaxSignatureAttempts expects a positive number, got %d", opts.MaxSignatureAttempts)}
+	if remoteOpts.MaxSignatureAttempts <= 0 {
+		return ocispec.Descriptor{}, verificationOutcomes, ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("verifyOptions.MaxSignatureAttempts expects a positive number, got %d", remoteOpts.MaxSignatureAttempts)}
 	}
```
We should perform this check at line 173 .

Reason: Why to resolve a reference when we are going to throw an error.

shizhMSFT reviewed (approved):
LGTM

shizhMSFT commented on notation.go:
```diff
@@ -156,31 +170,38 @@ func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, op
-		return ocispec.Descriptor{}, verificationOutcomes, ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("verifyOptions.MaxSignatureAttempts expects a positive number, got %d", opts.MaxSignatureAttempts)}
-	}
-	errExceededMaxVerificationLimit := ErrorVerificationFailed{Msg: fmt.Sprintf("total number of signatures associated with an artifact should be less than: %d", opts.MaxSignatureAttempts)}
+	errExceededMaxVerificationLimit := ErrorVerificationFailed{Msg: fmt.Sprintf("total number of signatures associated with an artifact should be less than: %d", remoteOpts.MaxSignatureAttempts)}
```
nit: we can merge this line to line 223

patrickzheng200 replied:
And we can move line 222 to line 193, i.e. use the return to replace the break statement in the for loop.

shizhMSFT replied:
The change related to this comment is reverted.

priteshbandi reviewed (approved):
LGTM

shizhMSFT reviewed (approved):
LGTM
